manager/state/raft/transport/peer.go

```diff
@@ -29,31 +29,58 @@ type peer struct {
 	cancel context.CancelFunc
 	done   chan struct{}
 
-	mu      sync.Mutex
-	cc      *grpc.ClientConn
-	addr    string
-	newAddr string
+	mu             sync.Mutex
+	cc             *grpc.ClientConn
+	ccAddr         string
+	alternateAddrs map[string]struct{}
 
 	active       bool
 	becameActive time.Time
 }
 
-func newPeer(id uint64, addr string, tr *Transport) (*peer, error) {
-	cc, err := tr.dial(addr)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to create conn for %x with addr %s", id, addr)
+func toAddrMap(curAddr string, addrs []string) map[string]struct{} {
+	addrMap := make(map[string]struct{}, len(addrs)-1)
+	for _, a := range addrs {
+		if a == curAddr {
+			continue
+		}
+		addrMap[a] = struct{}{}
+	}
+	return addrMap
+}
+
+func newPeer(id uint64, tr *Transport, addrs ...string) (*peer, error) {
+	if len(addrs) == 0 {
+		return nil, errors.New("addrs list can't be empty")
+	}
+	var (
+		cc   *grpc.ClientConn
+		addr string
+	)
+	for _, a := range addrs {
+		c, err := tr.dial(a)
+		if err == nil {
+			addr = a
+			cc = c
+			break
+		}
+		log.G(tr.ctx).WithError(err).Errorf("failed to create conn for %x with addr %s", id, a)
+	}
+	if cc == nil {
+		return nil, errors.New("dial failed for all provided addrs")
 	}
```

Comment:
I still think this loop should be moved into a function that returns (cc, addr).